action.yml

```diff
@@ -77,10 +77,149 @@ runs:
+            $RE = "[{0}]" -f [RegEx]::Escape($InvalidCharacters)
+            $FileName = [IO.Path]::GetFileName($URL) -Replace $RE
+
+            if ([String]::IsNullOrEmpty($FileName)) {
+              $FileName = [System.IO.Path]::GetRandomFileName()
+            }
+            $Path = Join-Path -Path "${env:Temp}" -ChildPath $FileName
+          }
+
+          Write-Host "Downloading package from $URL to $Path..."
+
+          $Interval = 30
+          $StartTime = Get-Date
+          do {
+            try {
+              $AttemptStartTime = Get-Date
+              (New-Object System.Net.WebClient).DownloadFile($URL, $Path)
+              $AttemptDuration = [math]::Round(($(Get-Date) - $AttemptStartTime).TotalSeconds, 2)
+              Write-Host "Package downloaded in $AttemptDuration seconds"
+              break
+            } catch {
+              $AttemptDuration = [math]::Round(($(Get-Date) - $AttemptStartTime).TotalSeconds, 2)
+              Write-Warning "Package download failed after $AttemptDuration seconds"
+              Write-Warning $_.Exception.Message
+            }
+
+            if ($Retries -eq 1) {
+              $Duration = [math]::Round(($(Get-Date) - $StartTime).TotalSeconds, 2)
+              throw "Package download failed after $Duration seconds"
+            }
+
+            Write-Warning "Waiting $Interval seconds before retrying (retries remaining: $Retries)..."
+            Start-Sleep -Seconds $Interval
+          } while (--$Retries -gt 0)
+
+          return $Path
+        }
+
+        function Verify-Checksum {
+          Param
+          (
+            [Parameter(Mandatory = $true)]
+            [String]$Actual,
+            [Parameter(Mandatory = $true)]
+            [String]$Expected
+          )
+
+          Write-Verbose "Performing Checksum Verification"
+          if ($Actual -eq $Expected) {
+            Write-Verbose "Checksum verification passed"
+          } else {
+            throw "Checksum verification failure (Actual: $Actual vs Expected: $Expected)"
+          }
+        }
+
+        function Invoke-Installation {
+          Param
+          (
+            [Parameter(Mandatory, ParameterSetName = "URL")]
+            [String] $URL,
+            [Parameter(Mandatory, ParameterSetName = "LocalPath")]
+            [String] $LocalPath,
+            [ValidateSet("MSI", "EXE")]
+            [String] $Type,
+            [String[]] $InstallArgs,
+            [String[]] $ExtraInstallArgs,
+            [String] $ExpectedSHA256
+          )
+
+          if ($PSCmdlet.ParameterSetName -eq "LocalPath") {
+            if (-not (Test-Path -Path $LocalPath)) {
+              throw "LocalPath parameter is specified, but the file does not exist"
+            }
+            $FilePath = $LocalPath
+          } else {
+            $FileName = [System.IO.Path]::GetFileName($URL)
+            $FilePath = Invoke-Download -URL $URL -Path "${env:Temp}\$FileName"
```

Comment:
`Invoke-Download` does the temp file dance as well as handling invalid characters.  Why don't you just pass the URL and let `Invoke-Download` do this work?

Reply:
I like that idea, done.